### PR TITLE
Next tick parameters

### PR DESCRIPTION
**Pull request by benjamingr**
Lesson learned, test added, code fixed. 

Fixes https://github.com/sinonjs/lolex/issues/122

### CODE REVIEW

File: test/lolex-test.js [hunk 1]
```diff
@@ -2053,6 +2053,17 @@ describe("lolex", function () {
                 clock.runAll();
                 assert(!called);
             });
+
+            it("passes arguments when installed - GitHub#122", function () {
+                var clock = this.clock = lolex.install();
```
Comment (fatso83) [inline test/lolex-test.js]: why assign to `this.clock`? It is never used in the code. It seems quite sufficient to only set the local variable.

Comment (benjamingr) [inline test/lolex-test.js]: Oh, I meant to do that and use it to uninstall - and then uninstalled myself. Fixing.

Comment (benjamingr) [inline test/lolex-test.js]: Fixed (removed uninstall call)

Comment (elad-nach) [inline test/lolex-test.js]: i think he meant on line 2058
`var clock = this.clock = lolex.install();`
could be changed to:
`var clock = lolex.install();`

Comment (benjamingr) [inline test/lolex-test.js]: @elad-nach - the problem was it also `clock.uninstall()`d, `var clock` is so I don't have to type `this` before it and there is a `afterEach` hook that uninstalls already. The tests above this one show the pattern.

**Comment by elad-nach** [comment]
this is somehow still broken.

I'm trying to integrate 2.1.1 to sinon and the nextTick functionality somehow is still breaking something else on the `test-headless`.

I'm trying to find out what exactly. 

Commenting it out the relevant nextTick lines resolves the problem.

**Comment by fatso83** [comment]
@elad-nach Using Sinon as an integration test is a pretty good idea in any case, which we perhaps might consider in any case. Thanks for doing it and uncovering stuff like this in any case. I'll try to do it myself the next time a patch comes in.

**Comment by benjamingr** [comment]
> I'm trying to integrate 2.1.1 to sinon and the nextTick functionality somehow is still breaking something else on the test-headless.

We can detect if the nextTick was run from Node code and not our own and run node's own microticks - can you try adding a line to run the original nextTick if the call originated in Node and see if that helps (by checking the current stack trace).

**Comment by elad-nach** [comment]
i fail to understand - what's the difference? nextTick is a node functionality.
and a couple of questions at that:
1. how was process.nextTick() handled before this change?
2. looking at the implementation - runJobs is equivalent to nodeJS' nextTick()? nextTick means that the callbacks should be executed after the user-code has finished executing - how do you defer the execution to that stage?

**Comment by benjamingr** [comment]
> how was process.nextTick() handled before this change? would it just hang?

Lolex would not handle `nextTick` at all - it would be like an "uncovered timer"

> looking at the implementation - runJobs is equivalent to nodeJS' nextTick()? 

No, `nextTick` is, `runJobs` is equivalent to "RunJobs" from here: http://ecma-international.org/ecma-262/8.0/#sec-runjobs

Node does this here:

 - First, a NextTickQueue is defined (I used an array for this here, since it's a testing library and a fast queue dependency isn't worth it). https://github.com/nodejs/node/blob/master/lib/internal/process/next_tick.js#L13
 - Doing a `process.nextTick` does https://github.com/nodejs/node/blob/master/lib/internal/process/next_tick.js#L107-L117 (pushes into the queue, albeit cleverly).
 - Which in turn has a hook in node.cc which runs all the microtasks after a tick. Here: https://github.com/nodejs/node/blob/master/src/node.cc#L1230-L1253 - as `runMicrotasks`. 
 - This hooks `RunMicrotasks` which is a V8 feature: https://github.com/nodejs/node/blob/master/src/node.cc#L1358 

> nextTick means that the callbacks should be executed after the user-code has finished executing - how do you defer the execution to that stage?

Same way as timers - I don't run them at all - but any timer ticking does. The only "special" thing is that I have to runJobs between any two opportunity for timers - since microticks `nextTick`d during a timer have to happen before the following timer (except in a specific edge case I'd rather not get into). 

```js
var clock = lolex.install();
process.nextTick(() => console.log("A"));
clock.runAll(); // this runs the nextTick, like a timer. 
```

Basically, the important part is that like other timers in lolex - we are able to tell it "progress asynchronous tasks".

**Comment by elad-nach** [comment]
as far as I understand - nextTick has nothing to do with time ticking; why invoke it from within the timer context when it has to satisfy other event loop state constraints.

Before the nextTick implementation on lolex - process.nextTick() was handled by node correctly:
```
describe('a test', function () {
    'use strict';
    const llx = require('lolex');
    it('should do the next tick', (done) => {
        llx.install();

        process.nextTick(() => {
            console.log('1');
            done();
        });
    })
})
```
this gets executed with lolex 1.6.0 and node 6.9.1 so I guess I'm missing something?

**Comment by benjamingr** [comment]
@elad-nach the whole point of lolex is to convert test code that is supposed to be asynchronous into more controllable synchronous code - however: it failed with code using nextTick - not only can the above code sample be written completely synchronously in lolex 2 - but requires an asynchronous test in lolex 1. 

Not to mention things that never worked (like timers inside `nextTick`). 

`nextTick` isn't about "time ticking" - but it has to happen in order to scheduler code relying on timers correctly. Once I get the one other edge case (promises) - we can write tests for complex asynchronous code synchronously with lolex - which would be great.

(setImmediate doesn't have anything to do with time ticking either for what it's worth)

**Comment by elad-nach** [comment]
timers inside `nextTick` wouldn't execute because timers have to be triggered as an interrupt, which was not possible when using lolex without utilising the original system timers for interrupts (that's why I implemented `shouldAdvanceTime` feature that piggybacks over the system timers' advancements), not because of `nextTick`'s behaviour.

Basically without explicitly calling `clock.runAll()` or `shouldAdvanceTime=true` (default is `false`) or calling `tick()` this implementation will never get called.

Anyone that will want to use a fake timer in an environment where `process.nextTick()` is called, will now have to manually `tick()` it, call the `runAll` or install with `shouldAdvanceTime=true`.

**Comment by benjamingr** [comment]
> timers inside nextTick wouldn't execute because timers have to be triggered as an interrupt, which was not possible when using lolex without utilising the original system timers for interrupts 

Timers inside `nextTick` wouldn't execute because `nextTick` wouldn't happen synchronously when you choose - but it was left to the platform (unlike things like setImmediate). 

> Basically without explicitly calling clock.runAll() or shouldAdvanceTime=true (default is false) or calling tick() this implementation will never get called.

Exactly, that's the point, you get to synchronize your flow.

>  Anyone that will want to use a fake timer in an environment where process.nextTick() is called, will now have to manually tick() it, call the runAll or install with shouldAdvanceTime=true.

Anyone who wants to use a fake timer which defaults to not progressing time - would not expect time to progress. Calling `nextTick` is _exactly_ like calling `setImmediate` (sans i/o order) and it made no sense to support the one and not another.

The point of lolex is to fake timers and to be able to break application tests into discrete synchronization points. Without nextTick - that never really worked for node (but worked fine for browsers).

What we _can_ do however is still run `nextTick` calls that are not a part of user code (like asynchronous code).

**Comment by elad-nach** [comment]
I agree with you on some of the points you've made but under real life conditions unless you execute every `nextTick` call synchronously (which might break other things for some users due to execution context) you won't be able to know when to `runAll` from within the test context because you don't know when `nextTick` is called.

Now what happens to sinon users that used the fake timers? will they all have to revise their tests as of such?

**Comment by benjamingr** [comment]
> you don't know when nextTick is called.

I think you can - it's called when you run your clock or tick it (we override nextTick on `.install`) - that was the whole point of the PR :)
